alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/nodes/AbstractNode.kt

```diff
@@ -0,0 +1,138 @@
+import java.util.Objects
+import java.util.Spliterator
+import java.util.concurrent.Semaphore
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.function.Consumer
+import javax.annotation.Nonnull
+
+/**
+ * This class realizes an abstract node. You may extend it to realize your own
+ * nodes.
+ *
+ * @param <T> concentration type
+</T> */
+abstract class AbstractNode<T> @JvmOverloads constructor(
+    /**
+     * The environment in which the node is places.
+     */
+    val environment: Environment<*, *>,
+    override val id: Int = idFromEnv(environment),
+    override val reactions: MutableList<Reaction<T>> = ArrayList(),
+    /**
+     * The node's molecules.
+     */
+    val molecules: MutableMap<Molecule, T> = LinkedHashMap(),
+    override val capabilities: MutableList<NodeProperty<T>> = ArrayList()
+) : Node<T> {
+
+    final override fun addReaction(reactionToAdd: Reaction<T>) {
+        reactions.add(reactionToAdd)
+    }
+
+    /**
+     * Default implementation fails: override correctly calling the constructor.
+     */
+    override fun cloneNode(currentTime: Time): Node<T> =
+        TODO("Cloning not yet implemented in ${this::class.simpleName}")
+
+    final override fun compareTo(@Nonnull other: Node<T>): Int = when (other) {
+        is AbstractNode<*> -> (id - other.id).coerceIn(-1, 1)
+        else -> 0
+    }
+
+    override fun contains(molecule: Molecule): Boolean = molecules.containsKey(molecule)
+
+    /**
+     * @return an empty concentration
+     */
+    protected abstract fun createT(): T
+
+    final override fun equals(other: Any?): Boolean = other is Node<*> && other.id == id
+
+    /**
+     * Performs an [action] for every reaction.
+     */
+    final override fun forEach(action: Consumer<in Reaction<T>>) = reactions.forEach(action)
+
+    override fun getConcentration(molecule: Molecule): T = molecules[molecule] ?: createT()
+
+    override val contents: Map<Molecule, T> = Collections.unmodifiableMap(molecules)
+
+    override val moleculeCount: Int
+        get() = molecules.size
+
+    final override fun hashCode(): Int {
+        return id // TODO: better hashing
+    }
+
+    final override fun iterator(): Iterator<Reaction<T>> {
+        return reactions.iterator()
+    }
+
+    final override fun removeConcentration(moleculeToRemove: Molecule) {
+        if (molecules.remove(moleculeToRemove) == null) {
+            throw NoSuchElementException("$moleculeToRemove was not present in node $id")
+        }
+    }
+
+    final override fun removeReaction(reactionToRemove: Reaction<T>) {
+        reactions.remove(reactionToRemove)
+    }
+
+    override fun setConcentration(molecule: Molecule, concentration: T) {
+        molecules[molecule] = concentration
+    }
+
+    final override fun addProperty(nodeProperty: NodeProperty<T>) {
+        capabilities.also {
+            it.removeIf { c -> c::class == nodeProperty::class }
+        }.add(nodeProperty)
```

Comment:
I do not understand what is going on here

Reply:
I thought that a node should not contain more than one property of the same type. I preferred removing any previously added property instead of throwing an exception both because it was convenient with the previous node architecture but also i would think a user is more likely to be interest in using the last property it inserted.

Reply:
It's way better to fail as fast and clearly as possible